### PR TITLE
Optimize zfs diff, round 2

### PR DESCRIPTION
**Pull request by rincebrain**
### Motivation and Context
I'd like to get `zfs diff` results before the sun swells into a red giant.

(What's the correct way to credit this? I'll obviously squash this before merge, but I don't know if I should do something beyond Signed-off-by on top...)

### Description
Originally based on @terem42's patch based on #10391, for much faster `zfs diff`.

I recommend reading #10391, since this is basically just build fixes - actually, I'm just going to quote it here.

> Motivation and Context
> zfs diff was found to be extremely slow to the point of being almost unusable. These changes often improve performance 10 - 100x.
> 
> The current implementation of zfs diff uses zfs_obj_to_stats(), which in turn uses zfs_obj_to_path() to get the path to the dnode. The path is then built using zap_value_search() to expand the parents. The problem is zap_value_search() doesn't do a lookup, it does a linear search through the entire directory object, which leads to n^2 behavior.
> 
> Description
> This changeset adds the ioctl ZFS_IOC_DUMP_ZAP, which is used to enumerate directories and cache them in userland. The zfs diff stream has been modified to always return a stat object for every id being used since the libzfs_diff code would always do this anyway. A new flag was added to zfs_obj_to_stats() to prevent looking up the path the old way.
> 
> These changes are intended to produce identical results to the current diff and thus do not fix any of the problems with the current version (e.g., any of the issues with hard links).

### How Has This Been Tested?
Well, first, I tried using kernel modules from this PR with normal git `zfs diff`, to see what would happen:
`Unable to obtain diffs: Unknown error 1031`

Which seems to be `ZFS_ERR_IOC_CMD_UNAVAIL`, and I'm kind of surprised we don't pretty print something better here...but it's not core dumping.

(I looked at plumbing in the old diff for backward compat, but it would essentially involve keeping all the pre-changed bits around and maintaining both, and I don't personally think it's worth it just to avoid `zfs diff` not returning an error with mismatched userland, of all things. If people disagree, we can go down that path, but that's my $0.02.)

I took a dataset I had laying around that's had a bunch of incremental churn and snapshots over a long period (my maildir), which reports the following on the live version:
```
    Dnode slots:
        Total used:       2755731
        Max used:         8960405
        Percent empty:  69.245464
```
`zfs send -R [latest] | zfs recv [single flat file pool]` and `zfs diff`ed a snapshot from 2020-12-20 with a snapshot from 2021-11-18 with the old and new code (on copies of the identical pool file, to be clear).

According to `time`, 2.0.6's `zfs diff` took:
```
2.01user 4053.73system 1:07:56elapsed 99%CPU (0avgtext+0avgdata 5080maxresident)k
128inputs+23032outputs (0major+303minor)pagefaults 0swaps
```

while this code took:
```
5.35user 10.21system 0:10.49elapsed 148%CPU (0avgtext+0avgdata 585448maxresident)k
312inputs+23032outputs (4major+145374minor)pagefaults 0swaps
```

So 1h7m to 10.5s. Not bad.

Other than that, just the test suite. (I'm somewhat curious to try this on the dataset I was using for #12072, but that's on a machine I don't really want to try new code on, and over a TB, so slightly cumbersome to pass around...)

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by rincebrain** [comment]
Based on [this test run](https://github.com/openzfs/zfs/runs/4490458512?) and [this test run](https://github.com/openzfs/zfs/runs/4490626484) having so much overlap in test failures, and on having looked and found one of them was because the CI bumped the kernel to a version post "cannot write to /dev/null from kernel", I'm going to suspect Ubuntu bumping to 5.11.x is the reason 20.04's CI runs are failing r/n.

File: include/sys/zfs_ioctl.h [hunk 1]
```diff
@@ -386,20 +387,27 @@ typedef struct dmu_replay_record {
 
 /* diff record range types */
 typedef enum diff_type {
-	DDR_NONE = 0x1,
-	DDR_INUSE = 0x2,
-	DDR_FREE = 0x4
+	DDR_NONE	= 0x0,
+	DDR_IN_TO	= 0x1,	/* dnode in to snap only */
+	DDR_IN_FROM	= 0x2,	/* dnode in from snap only */
+	DDR_IN_BOTH	= 0x3	/* dnode in both from and to snaps */
```
Comment (nabijaczleweli) [inline include/sys/zfs_ioctl.h:393]: `= D_I_T | D_I_F`?

File: module/zfs/zfs_ioctl.c [hunk 2]
```diff
@@ -1089,13 +1089,34 @@ zfs_secpolicy_config(zfs_cmd_t *zc, nvlist_t *innvl, cred_t *cr)
...
-	error = zfs_secpolicy_write_perms(zc->zc_name, ZFS_DELEG_PERM_DIFF, cr);
-	return (error);
+	err = zfs_secpolicy_write_perms(zc->zc_name, ZFS_DELEG_PERM_DIFF, cr);
+	return (err);
+}
+
+/*
+ * Policy for object to name lookups (both from and to datasets).
+ */
+/* ARGSUSED */
```
Comment (nabijaczleweli) [inline module/zfs/zfs_ioctl.c:1104]: sus :0

Comment (rincebrain) [reply module/zfs/zfs_ioctl.c]: LINE 1104 was SUS, 0 (new) ARGSUSED remain.

File: include/sys/zfs_ioctl.h [hunk 3]
```diff
@@ -386,20 +387,27 @@ typedef struct dmu_replay_record {
...
  * The diff reports back ranges of free or in-use objects.
  */
 typedef struct dmu_diff_record {
-	uint64_t ddr_type;
-	uint64_t ddr_first;
-	uint64_t ddr_last;
+	diff_type_t	ddr_type;
+	uint64_t	ddr_obj;
+	int		ddr_err[2];
+	zfs_diff_stat_t	ddr_zds[2];	/* space for 1 or two stats as needed */
```
Comment (nabijaczleweli) [inline include/sys/zfs_ioctl.h:403]: "1" or.. "two"? also, isn't this fundamentally what a two-element array, uh, is?

Comment (rincebrain) [reply include/sys/zfs_ioctl.h:403]: I presume they were trying to communicate that it's not always expected for both to be populated.

**Comment by jumbi77** [comment]
@rincebrain This looks really promising. Do you may can rebase it and ping again some reviewers? Would be great to get this improvement. Much thanks

**Comment by jumbi77** [comment]
@rincebrain sorry to respond to this closed PR, but can I may ask why this was unmerged closed? Your original description looks promising and sounds like a nice improvement.

And, most important: Thanks for your contribution in any case :)

**Comment by jumbi77** [comment]
In the predecessor PR #10391 the author of this PR (sadly) said that the closed this nice optimization because of lack of feedback. Can I may ask you @rincebrain to rebase this? Since some time has passed now and I got feeling that @tonyhutter, @behlendorf, @robn and other maintainers recently trying to give as much feedback as time allows, there is a good chance that it will be integrated.

**Comment by rincebrain** [comment]
You're welcome to open your own PR if you want.

**Pull request closed without merging**